### PR TITLE
Add ImageManager class

### PR DESCRIPTION
**Pull request by djberg96**
This PR adds a VirtualMachineImageManager class to handle methods for dealing with VM images. With this class you can do:

```
vmim = VirtualMachineImageManager.new
p vmim.publishers('eastus')
p vmim.offers('eastus', 'Canonical')
p vmim.skus('UbuntuServer', 'eastus', 'Canonical')
p vmim.versions('14.04.2', 'UbuntuServer', 'eastus', 'Canonical')
```

This does cause some breakage, as the publishers method was moved out of the base class and the offers method was moved out of the VirtualMachineManager class. However, I feel this is a better organization conceptually, and it more closely resembles both the Azure CLI and the old ruby sdk for Azure in terms of organization.

I will eventually add a list method that collates this information in the same manner in which the CLI command "azure vm image list" does.

### CODE REVIEW

**Comment by bzwei** [comment]
Please rebase -i to consolidate multiple commits.

**Comment by djberg96** [comment]
Why? What's the advantage of squashing commits?

**Comment by bzwei** [comment]
I think we prefer not too many tiny commits for one issue, for example, all the commits in this particular PR are tightly related for introducing the new image manager. @jrafanie and @Fryguy please confirm.

File: lib/azure/armrest/virtual_machine_image_manager.rb [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      #
+      #   vmim.offers('eastus', 'Canonical')
+      #
+      def offers(location = @location, publisher = @publisher)
+        raise ArgumentError, "No location specified" unless location
+        raise ArgumentError, "No publisher specified" unless publisher
+
+        url = build_url(location, 'publishers', publisher, 'artifacttypes', 'vmimage', 'offers')
+
+        JSON.parse(rest_get(url)).map{ |element| element['name'] }
```
Comment (bzwei) [inline lib/azure/armrest/virtual_machine_image_manager.rb]: Shouldn't this be `JSON.parse(rest_get(url))['value'].map { ... }`? Same in a few other places here.

Comment (djberg96) [inline lib/azure/armrest/virtual_machine_image_manager.rb]: @bzwei, no, I double checked. There's no 'value' key in the returned json results. In the offers method, for example, the result of JSON.parse(rest_get(url)) is:

```[{"location"=>"eastus", "name"=>"Ubuntu15.04Snappy", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/Ubuntu15.04Snappy"}, {"location"=>"eastus", "name"=>"Ubuntu15.04SnappyDocker", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/Ubuntu15.04SnappyDocker"}, {"location"=>"eastus", "name"=>"UbunturollingSnappy", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/UbunturollingSnappy"}, {"location"=>"eastus", "name"=>"UbuntuServer", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/UbuntuServer"}]
[{"location"=>"eastus", "name"=>"Ubuntu15.04Snappy", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/Ubuntu15.04Snappy"}, {"location"=>"eastus", "name"=>"Ubuntu15.04SnappyDocker", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/Ubuntu15.04SnappyDocker"}, {"location"=>"eastus", "name"=>"UbunturollingSnappy", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/UbunturollingSnappy"}, {"location"=>"eastus", "name"=>"UbuntuServer", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/UbuntuServer"}]```

**Comment by djberg96** [comment]
Ok, rebased the commits.

**Comment by jrafanie** [comment]
@djberg96 @bzwei My personal preference is to see small commits during development, I will then try to squash them so each commit has a single sentence subject that does one thing.  I try to avoid using `and` to join to different things in the same commit.  There is no rule about it.  I like to ask myself if separate git hunks relate to the same subject, if not, I'll split them.

To be totally honest, the only time squashing commits is really required is to squash a "try this commit" with "oops fix typo" commit.  I would always prefer smaller commits over enormous ones that do too many unrelated things.  You can always squash commits later.

There's advice all over the internet like [this](https://secure.phabricator.com/book/phabflavor/article/writing_reviewable_code/) that I generally like.

**Comment by djberg96** [comment]
@jrafanie Thanks.

@bzwei and @blomquisg I'd like to get this merge and then release it as 0.0.1, perhaps releasing it as a pre-release gem.

**Comment by blomquisg** [comment]
:+1:

**Pull request merged**